src/oscore/oscore_crypto.c

```diff
@@ -0,0 +1,161 @@
+/* -*- Mode: C; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+
+/*
+ * Copyright (c) 2018, SICS, RISE AB
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+/**
+ * @file oscore_crypto.c
+ * @brief An implementation of the Hash Based Key Derivation Function (RFC) and
+ * wrappers for AES-CCM*.
+ *
+ * \author
+ *      Martin Gunnarsson  <martin.gunnarsson@ri.se>
+ * extended for libcoap
+ *      Peter van der Stok <consultancy@vanderstok.org>
+ *      on request of Fairhair alliance
+ * adapted for libcoap integration
+ *      Jon Shallow <supjps-libcoap@jpshallow.com>
+ */
+
+#include "coap3/coap_internal.h"
+#include <string.h>
+
+#include <stdio.h>
+
+/*
+ * return 1 fail
+ *        0 OK
+ */
+int
+oscore_hmac_shaX(cose_alg_t hkdf_alg,
+                 coap_bin_const_t *key,
+                 coap_bin_const_t *data,
+                 coap_bin_const_t **hmac) {
+  if (!coap_crypto_hmac(hkdf_alg, key, data, hmac)) {
+    coap_log(LOG_WARNING, "hmac_shaX: Failed hmac\n");
+    return 1;
+  }
+  return 0;
+}
+
+/*
+ * return 1 fail
+ *        0 OK
+ */
+int
+oscore_hkdf_extract(cose_alg_t hkdf_alg,
+                    coap_bin_const_t *salt,
+                    coap_bin_const_t *ikm,
+                    coap_bin_const_t **hkdf_extract) {
+  assert(ikm);
+  if (salt == NULL || salt->s == NULL) {
+    uint8_t zeroes_data[32];
+    coap_bin_const_t zeroes;
+
+    memset(zeroes_data, 0, 32);
```

Comment:
For safety reasons, I suggest using `sizeof(zeroes_data)` here and for `zeroes.length`.

Reply:
Agreed and corrected.